sfn/sfn.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SFN" "1" "February 2013" "" ""
.
.SH "NAME"
\fBsfn\fR \- send files over network
.
.SH "SYNOPSIS"
\fBsfn\fR \fB\-\-listen\fR [options] [files to send]
.
.br
\fBsfn\fR \fB\-\-connect\fR <address> [options] [files to send]
.
.SH "DESCRIPTION"
sfn is an utility to send files over network using direct TCP connection\. After launched, sfn establishes a connection, sends all the files from command line (if any), receives all the files from another side and then exits\.
.
.SH "OPTIONS"
.
.TP
\fB\-\-version\fR, \fB\-v\fR
Show version and exit\.
.
.TP
\fB\-\-help\fR, \fB\-h\fR
Show help and exit\.
.
.TP
\fB\-\-port\fR, \fB\-p\fR
Use specified port\. Defaults to 3214\.
.
.TP
\fB\-\-prefix\fR, \fB\-f\fR
Add prefix to received files\' path and name\. For example: \fB/home/user/downloads/\fR, \fBsfn\-\fR, \fB/etc/file\-\fR\.
.
.TP
\fB\-\-no\-external\-ip\fR, \fB\-n\fR
Don\'t perform external IP detection and reverse DNS lookup\.
.
.SH "EXAMPLES"
Listen on port 3333, place all received files into current workdir:
.
.IP "" 4
.
.nf

$ sfn \-s \-p 3333
.
.fi
.
.IP "" 0
.
.P
Connect to server \fB55\.66\.77\.88\fR, send a file named \fBnumbers\.txt\fR and place received ones into \fB/tmp/sfn\fR:
.
.IP "" 4
.
.nf

$ sfn \-c 55\.66\.77\.88 \-f /tmp/sfn/ numbers\.txt
.
.fi
.
.IP "" 0
.
.P
Please note that trailing slash in \fB/tmp/sfn/\fR is mandatory because it is simply a prefix\.
.
.P
Listen on default port, send some files and don\'t try to determine our external IP:
.
.IP "" 4
.
.nf

$ sfn \-s \-n for\-james\-*\.txt
.
.fi
.
.IP "" 0
.
.P
Connect to \fBmyserver\.dyndns\.org\fR, receive some files, prefix their names with \fBsfn\-received\-\fR and place them into \fB~/Downloads\fR:
.
.IP "" 4
.
.nf

$ sfn \-c myserver\.dyndns\.org \-f ~/Downloads/sfn\-received\-
.
.fi
.
.IP "" 0
.
.SH "BUGS"
If you have found a bug, create an issue in the project\'s issue tracker: \fIhttps://github\.com/m1kc/sfn/issues\fR or simply drop me a note on email\. Thanks\.
.
.SH "COPYRIGHT"
Copyright (c) m1kc <\fIm1kc@yandex\.ru\fR>, 2013


ChangeLog
2011-05-03 Majki <majki@majki.hu>
	* 2.0.1-8 :
	notify_notification_new fix

	2010-02-19 Majki <majki@majki.hu>
	* 2.0.1-7 :
	Fixed network manager library problem (renamed lib file)
	  ^- Thanks to leprechau

2009-06-23 Majki <majki@majki.hu>
	* 2.0.1-6 :
	Fixed duplicated dependency sharutils
	Fixed .desktop file
	Fixed icon update after install
	All fixes thx to alessandro_ufms

2009-06-16 Majki <majki@majki.hu>
	* 2.0.1-5 :
	Added sharutils as make dependency (thx to tschertel)

2009-04-07 Majki <majki@majki.hu>

	* 2.0.1-4 :
	compile error, libgnomeui-2.0 bug
	patched configure script (thx to solskogen)
	
2009-03-12 Majki <majki@majki.hu>

	* 2.0.1-3 :
	pkgconfig added as make dependency (thx to cecampos)

2009-02-20 Majki <majki@majki.hu>

	* 2.0.1-2 :
	Fixed Sourceforce download url

2009-02-20 Majki <majki@majki.hu>

	* 2.0.1-1 :
	Initial PKGBUILD



sfn/sfn.d
/*
 * sfn.d
 *
 * Copyright 2012 m1kc <m1kc@yandex.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 *
 */

import std.stdio;
import std.file;
import std.socket;
import std.socketstream;
import std.getopt;
import std.conv;
import std.string;
import std.path;
import std.datetime;
import std.format;
import std.array;
import core.thread;
import std.compiler;

__gshared bool localDone = false;
__gshared bool remoteDone = false;
__gshared SocketStream stream = null;
__gshared Socket listener = null;
__gshared Socket socket = null;
__gshared string[] send;
__gshared string prefix = "";

immutable uint windowSize = 1024*64;

//version = TestBarAndExit;
//version = TestNumbersAndExit;
void main(string[] args)
{
	version(TestBarAndExit)
	{
		long time = currentTime();
		int max = int.max/10000;
		for (int i=0; i<max; i++) showBar(i, max, time);
		return;
	}
	version(TestNumbersAndExit)
	{
		int max = int.max;
		for (int i=0; i<max; i++) std.stdio.write(numberOfBytes(i), "\r");
		writeln;
		return;
	}

	bool server = false;
	string connect = null;
	ushort port = 3214;
	bool help = false;
	bool ver = false;
	bool noExtIP = false;

	getopt(args,
		"version|v", &ver,
		"help|h", &help,
		"listen|l|s", &server,
		"connect|c", &connect,
		"port|p", &port,
		"prefix|f", &prefix,
		"no-external-ip|n", &noExtIP,
	);
	send = args[1..$];

	if (help) { usage(); return; }
	if (ver) { writeln("sfn 1.15" ~ "\nCompiled by: " ~ std.compiler.name); return; }

	if ((connect is null) && !server) { usage("You must specify mode."); return; }
	if ((connect !is null) && server) { usage("You must specify only one mode."); return; }

	if (server)
	{
		listener = new TcpSocket();
		assert(listener.isAlive);
		listener.bind(new InternetAddress(port));
		listener.listen(10);
		writeln("Waiting for connection...");
		if (!noExtIP)
		{
			Thread ext = new Thread( &extIP );
			ext.start();
		}
		socket = listener.accept();
		writeln("Connected: " ~ to!string(socket.remoteAddress()));
		stream = new SocketStream(socket);
	}
	else
	{
		writeln("Connecting...");
		socket = new TcpSocket(new InternetAddress(connect, port));
		writeln("Connected.");
		stream = new SocketStream(socket);
	}
	scope(exit)
	{
		stream.close();
		socket.close();
		if (listener !is null) listener.close();
	}
	Thread receiver = new Thread( &receiveFiles );
	receiver.start();
	Thread sender = new Thread( &sendFiles );
	sender.start();
	writeln("Started transfer.");
	// wait
	while(!localDone || !remoteDone) Thread.sleep( dur!("msecs")(100) );
	writeln("Transfer complete.");
	return;
}

long currentTime()
{
	return Clock.currStdTime()/10_000; // hnsecs (hecto-nanoseconds (100 ns)) to millisecs
}

string numberOfBytes(long x)
{
	if (x<1024) return to!string(x) ~ " bytes";

	auto writer = std.array.appender!string();
	if (x<1024*1024) formattedWrite(writer, "%.1f Kb", cast(double)(x)/1024);
	else if (x<1024*1024*1024) formattedWrite(writer, "%.1f Mb", cast(double)(x)/(1024*1024));
	else formattedWrite(writer, "%.1f Gb", cast(double)(x)/(1024*1024*1024));
	return writer.data;
}

version(Windows) // issue #17
{
	int getTerminalWidth()
	{
		return 80;
	}
}
else
{
	// C extension (ioctl)
	extern (C) int getTerminalWidth();
}

void showBar(ulong progress, ulong total, long startTime = -1)
{
	string output = " ";

	long timeSpent = currentTime()-startTime;
	if (timeSpent != 0)
	{
		output ~= numberOfBytes(progress*1000/timeSpent);
		output ~= "/sec  ";
	}

	output ~= numberOfBytes(progress);
	output ~= "/";
	output ~= numberOfBytes(total);
	output ~= "\r";

	ulong terminalWidth = getTerminalWidth();
	ulong barsCount = terminalWidth
		-3								// [], space
		-output.length
		;
	ulong bars = progress*(barsCount+1)/total;
	write("[");
	for (ulong i=0; i<barsCount; i++) write(i<bars ? "#" : "-");
	write("] ");

	write(output);
}

void receiveFiles()
{
	immutable ubyte FILE = 0x01;
	immutable ubyte DONE = 0x02;

	ubyte b;

	while(true)
	{
		stream.read(b);
		if (b == DONE)
		{
			writeln("Remote done.");
			remoteDone = true;
			return;
		}
		else if (b == FILE)
		{
			write("Receiving a file: ");
			string filename = to!string(stream.readLine());
			write(filename ~ ", ");
			ulong size;
			stream.read(size);
			writeln(size, " bytes");

			File f = File(prefix ~ filename, "w");

			ubyte[] buf = new ubyte[windowSize];
			ulong remain = size;
			ulong readc;
			long time = currentTime();
			while(remain > 0)
			{
				if (remain < windowSize) buf = new ubyte[cast(uint)(remain)];
				readc = stream.read(buf);
				remain -= readc;
				f.rawWrite(buf[0..cast(uint)(readc)]);
				showBar(size-remain, size, time);
			}
			writeln();
			writeln("Done.");
			f.close();
		}
		else
		{
			writeln("Unexpected byte " ~ to!string(b));
		}
	}
}

void sendFiles()
{
	immutable ubyte FILE = 0x01;
	immutable ubyte DONE = 0x02;

	foreach(string s; send)
	{
		writeln("Sending a file: " ~ s);
		DirEntry d = dirEntry(s);
		if (d.isFile())
		{
			stream.write(FILE);
			File f = File(s, "r");
			stream.writeString(f.name.split(dirSeparator)[$-1]);
			stream.writeString("\n");
			stream.write(f.size());
			writeln(to!string(f.size()) ~ " bytes");
			ulong size = f.size();
			ulong sent = 0;
			long time = currentTime();
			foreach (ubyte[] b; f.byChunk(windowSize))
			{
				stream.write(b);
				sent += b.length;
				showBar(sent, size, time);
			}
			writeln();
			writeln("Done.");
			f.close();
		}
		else
		{
			writeln("Error: does not exist or is not a file");
		}
	}

	stream.write(DONE);
	localDone = true;
	writeln("Local done.");
}

void extIP()
{
	// URL: http://tomclaw.com/services/simple/getip.php

	Socket sock = new TcpSocket(new InternetAddress("tomclaw.com", 80));
	scope(exit) sock.close();
	SocketStream ss = new SocketStream(sock);

	ss.writeString(
		"GET /services/simple/getip.php HTTP/1.0\r\n"
		"Host: tomclaw.com\r\n"
		"\r\n"
	);

	// headers
	while(true)
	{
		auto line = ss.readLine();
		if (line.length==0) break;
	}

	// IP
	auto ip = ss.readLine();
	writeln("External IP: ", ip);

	// reverse DNS lookup
	InternetHost ih = new InternetHost();
	if (ih.getHostByAddr(ip))
	{
		writeln("Host: ", ih.name);
		foreach (string s; ih.aliases) writeln("Alias: ", s);
	}
	else
	{
		writeln("Host: ", "can't determine");
	}
}

void usage(string error = null)
{
	if (error !is null) writeln(error); else write("sfn - send files over network. ");
	write("Usage:

    sfn --listen [options] [files to send]
    sfn --connect <address> [options] [files to send]

sfn will establish a connection, send all the files, receive all the files from another side and then exit.

-l and -s are aliases for --listen, -c is an alias for --connect.

Options:

    --version, -v         Show sfn version and exit.
    --help, -h            Show this text and exit.
    --port, -p            Use specified port. Defaults to 3214.
    --prefix, -f          Add prefix to received files' path and name. For example: '/home/user/downloads/', 'sfn-', '/etc/file-'.
    --no-external-ip, -n  Don't perform external IP detection and reverse DNS lookup.

");

}
